File: samples/C# Samples/WebcamDemo/WebcamDemo/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Dynamsoft.UVC;
using Dynamsoft.Core;
using System.IO;
using System.Runtime.InteropServices;
using Dynamsoft.Common;

namespace WebcamDemo
{
    public partial class Form1 : Form
    {
        private int m_iDesignWidth = 755;
        private int iControlWidth = 275;
        private int iControlHeight = 295;
        private CameraManager m_CameraManager = null;
        private ImageCore m_ImageCore = null;
        private string m_StrProductKey = "t0068UwAAAEQABDxqjGfgEzhVYureL0kGxugcsvIqCDGTPTsR5nLaQsNupIc17Y5vpMZAWBDsd6Xw3NMYzdHlHwiKUrfe/cU=";

        private Camera m_CurrentCamera = null;


        public Form1()
        {
            InitializeComponent();
            m_CameraManager = new CameraManager(m_StrProductKey);
            m_ImageCore = new ImageCore();
            dsViewer1.Bind(m_ImageCore);
            this.Load += new EventHandler(Form1_Load);
        }

        void Form1_Load(object sender, EventArgs e)
        {
            try
            {
                m_iDesignWidth = this.Width;
                if (m_CameraManager.GetCameraNames()!=null)
                {
                    foreach (string temp in m_CameraManager.GetCameraNames())
                    {
                        cbxSources.Items.Add(temp);
                    }

                    cbxSources.SelectedIndexChanged += cbxSources_SelectedIndexChanged;
                    if (cbxSources.Items.Count > 0)
                        cbxSources.SelectedIndex = 0;
                }
            }
            catch (Exception exp)
            {
                MessageBox.Show(exp.Message);
            }
        }

        private void btnRemoveAllImages_Click(object sender, EventArgs e)
        {
            m_ImageCore.ImageBuffer.RemoveAllImages();
        }

        private void btnCaptureImage_Click(object sender, EventArgs e)
        {
            if(m_CurrentCamera!=null)
            {
                Bitmap tempBmp = m_CurrentCamera.GrabImage();
                m_ImageCore.IO.LoadImage(tempBmp);
            }
        }

        private void cbxSources_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (m_CurrentCamera != null)
            {
                m_CurrentCamera.OnFrameCaptrue -= m_CurrentCamera_OnFrameCaptrue;
                m_CurrentCamera.Close();
            }
            if (m_CameraManager != null)
            {
                m_CurrentCamera = m_CameraManager.SelectCamera((short)cbxSources.SelectedIndex);
                m_CurrentCamera.Open();

                m_CurrentCamera.OnFrameCaptrue += m_CurrentCamera_OnFrameCaptrue;
                ResizePictureBox();
            }
            if (m_CurrentCamera != null)
            {
                cbxResolution.Items.Clear();
                foreach (CamResolution camR in m_CurrentCamera.SupportedResolutions)
                {
                    cbxResolution.Items.Add(camR.ToString());
                }
                cbxResolution.SelectedIndexChanged += cbxResolution_SelectedIndexChanged;
                if (cbxResolution.Items.Count > 0)
                    cbxResolution.SelectedIndex = 0;
                ResizePictureBox();
            }

        }
        private void cbxResolution_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (cbxResolution.Text != null)
            {
                string[] strWXH = cbxResolution.Text.Split(new char[] { ' ' });
                if (strWXH.Length == 3)
                {
                    try
                    {
                        m_CurrentCamera.CurrentResolution = new CamResolution(int.Parse(strWXH[0]), int.Parse(strWXH[2]));
                    }
                    catch { }
                }
                m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_0);
                ResizePictureBox();
            }
        }


        private void pictureBox1_MouseClick(object sender, MouseEventArgs e)
        {
            Point tempPoint = new Point(e.X,e.Y);
            float tempWidth = pictureBox1.Width;
            float tempHeight = pictureBox1.Height;

            float imageWidth = m_CurrentCamera.CurrentResolution.Width;
            float imageHeight = m_CurrentCamera.CurrentResolution.Height;
            float zoomX = imageWidth / tempWidth;
            float zoomY = imageHeight / tempHeight;
            Point tempFocusPoint = new Point((int)(e.X * imageWidth),(int)(e.Y*imageHeight));
            Rectangle tempRect = new Rectangle(tempFocusPoint,new Size(50,50));
            m_CurrentCamera.FocusOnArea(tempRect);
        }
        
        private void ResizePictureBox()
        {
            if (m_CurrentCamera != null)
            {
                CamResolution camResolution = m_CurrentCamera.CurrentResolution;
                if (camResolution != null && camResolution.Width > 0 && camResolution.Height > 0)
                {
                    {
                        int iVideoWidth = iControlWidth;
                        int iVideoHeight = iControlWidth * camResolution.Height / camResolution.Width;
                        int iContentHeight = panel1.Height - panel1.Margin.Top - panel1.Margin.Bottom - panel1.Padding.Top - panel1.Padding.Bottom;
                        if (iVideoHeight < iContentHeight)
                        {
                            pictureBox1.Location = new Point(0, (iContentHeight - iVideoHeight) / 2);
                            pictureBox1.Size = new Size(iVideoWidth, iVideoHeight);
                        }
                        else
                        {
                            pictureBox1.Location = new Point(0, 0);
                            pictureBox1.Size = new Size(pictureBox1.Width, pictureBox1.Height);
                        }
                    }
                }
            }
        }
        private void RotatePictureBox()
        {
            if (m_CurrentCamera != null)
            { 
                CamResolution camResolution = m_CurrentCamera.CurrentResolution;
                if (camResolution != null && camResolution.Width > 0 && camResolution.Height > 0)
                {
                    if (camResolution.Width / camResolution.Height > iControlWidth / iControlHeight)
                    {
                        int iVideoHeight = iControlHeight;
                        int iVideoWidth = iControlHeight * camResolution.Height / camResolution.Width;
                        int iContentWidth = panel1.Width - panel1.Margin.Left - panel1.Margin.Right - panel1.Padding.Left - panel1.Padding.Right;
                        pictureBox1.Location = new Point((iContentWidth - iVideoWidth) / 2, 0);
                        pictureBox1.Size = new Size(iVideoWidth, iVideoHeight);
                    }
                    else
                    {
                        int iVideoWidth = iControlWidth;
                        int iVideoHeight = iControlWidth * camResolution.Width / camResolution.Height;
                        int iContentHeight = panel1.Height - panel1.Margin.Top - panel1.Margin.Bottom - panel1.Padding.Top - panel1.Padding.Bottom;
                        pictureBox1.Location = new Point(0, (iContentHeight - iVideoHeight) / 2);
                        pictureBox1.Size = new Size(iVideoWidth, iVideoHeight);
                    }
                }
            }
        }
        private void SetPicture(Image img)
        {
            Bitmap temp = ((Bitmap)(img)).Clone(new Rectangle(0, 0, img.Width, img.Height), img.PixelFormat);
            if (pictureBox1.InvokeRequired)
            {
                pictureBox1.BeginInvoke(new MethodInvoker(
                delegate ()
                {
                    pictureBox1.Image = temp;
                }));
            }
            else
            {
                pictureBox1.Image = temp;
            }

        }
        private void picbox90_Click(object sender, EventArgs e)
        {
            m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_90);
            RotatePictureBox();
        }

        private void picbox180_Click(object sender, EventArgs e)
        {
            m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_180);
            ResizePictureBox();
        }

        private void picbox270_Click(object sender, EventArgs e)
        {
            m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_270);
            RotatePictureBox();
        }
        void m_CurrentCamera_OnFrameCaptrue(Bitmap bitmap)
        {
            SetPicture(bitmap);
        }
        private void picbox90_MouseHover(object sender, EventArgs e)
        {
            picbox90.Image = global::WebcamDemo.Properties.Resources._90_hover;
        }
        private void picbox90_MouseLeave(object sender, EventArgs e)
        {
            picbox90.Image = global::WebcamDemo.Properties.Resources._90_normal;
        }
        private void picbox180_MouseHover(object sender, EventArgs e)
        {
            picbox180.Image = global::WebcamDemo.Properties.Resources._180_hover;
        }
        private void picbox180_MouseLeave(object sender, EventArgs e)
        {
            picbox180.Image = global::WebcamDemo.Properties.Resources._180_normal;
        }
        private void picbox270_MouseHover(object sender, EventArgs e)
        {
            picbox270.Image = global::WebcamDemo.Properties.Resources._270_hover;
        }
        private void picbox270_MouseLeave(object sender, EventArgs e)
        {
            picbox270.Image = global::WebcamDemo.Properties.Resources._270_normal;
        }
        
    }
}

File: samples/VB .NET Samples/WebcamDemo/WebcamDemo/Form1.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Dynamsoft.UVC
Imports Dynamsoft.Core
Imports System.IO
Imports System.Runtime.InteropServices
Imports Dynamsoft.Common

Partial Public Class Form1
    Inherits Form
    Private m_iDesignWidth As Integer = 755
    Private m_CameraManager As CameraManager = Nothing
    Private m_ImageCore As ImageCore = Nothing
    Private m_StrProductKey As String
    Private iControlWidth As Integer = 275
    Private iControlHeight As Integer = 295
    Private m_CurrentCamera As Camera = Nothing


    Public Sub New()
        InitializeComponent()
        m_StrProductKey = "t0068UwAAAEQABDxqjGfgEzhVYureL0kGxugcsvIqCDGTPTsR5nLaQsNupIc17Y5vpMZAWBDsd6Xw3NMYzdHlHwiKUrfe/cU="
        m_CameraManager = New CameraManager(m_StrProductKey)
        m_ImageCore = New ImageCore()
        dsViewer1.Bind(m_ImageCore)
        AddHandler Me.Load, New EventHandler(AddressOf Form1_Load)
    End Sub



    Private Sub Form1_Load(sender As Object, e As EventArgs)
        Try
            m_iDesignWidth = Me.Width
            If m_CameraManager.GetCameraNames() IsNot Nothing Then
                For Each temp As String In m_CameraManager.GetCameraNames()
                    cbxSources.Items.Add(temp)
                Next

                AddHandler cbxSources.SelectedIndexChanged, AddressOf cbxSources_SelectedIndexChanged
                If cbxSources.Items.Count > 0 Then
                    cbxSources.SelectedIndex = 0
                End If
            End If
        Catch exp As Exception
            MessageBox.Show(exp.Message)
        End Try

    End Sub

    Private Sub btnRemoveAllImages_Click(sender As Object, e As EventArgs) Handles btnRemoveAllImages.Click
        m_ImageCore.ImageBuffer.RemoveAllImages()
    End Sub


    Private Sub cbxSources_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cbxSources.SelectedIndexChanged
        If m_CurrentCamera IsNot Nothing Then
            RemoveHandler Me.m_CurrentCamera.OnFrameCaptrue, New Dynamsoft.UVC.Delegate.OnFrameCaptureHandler(AddressOf m_CurrentCamera_OnFrameCaptrue)
            m_CurrentCamera.Close()
        End If

        If m_CameraManager IsNot Nothing Then
            m_CurrentCamera = m_CameraManager.SelectCamera(CShort(cbxSources.SelectedIndex))
            m_CurrentCamera.Open()
            AddHandler Me.m_CurrentCamera.OnFrameCaptrue, New Dynamsoft.UVC.Delegate.OnFrameCaptureHandler(AddressOf m_CurrentCamera_OnFrameCaptrue)
            ResizePictureBox()
        End If

        If m_CurrentCamera IsNot Nothing Then
            cbxResolution.Items.Clear()

            For Each camR As CamResolution In m_CurrentCamera.SupportedResolutions
                cbxResolution.Items.Add(camR.ToString())
            Next
            If cbxResolution.Items.Count > 0 Then cbxResolution.SelectedIndex = 0
            ResizePictureBox()
        End If
    End Sub
    Private Sub cbxResolution_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cbxResolution.SelectedIndexChanged
        If cbxResolution.Text IsNot Nothing Then
            Dim strWXH As String() = cbxResolution.Text.Split(New Char() {" "c})

            If strWXH.Length = 3 Then

                Try
                    m_CurrentCamera.CurrentResolution = New CamResolution(Integer.Parse(strWXH(0)), Integer.Parse(strWXH(2)))
                Catch
                End Try
            End If

            m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_0)
            ResizePictureBox()
        End If
    End Sub
    Private Sub RotatePictureBox()
        If m_CurrentCamera IsNot Nothing Then
            Dim camResolution As CamResolution = m_CurrentCamera.CurrentResolution

            If camResolution IsNot Nothing AndAlso camResolution.Width > 0 AndAlso camResolution.Height > 0 Then

                If camResolution.Width / camResolution.Height > iControlWidth / iControlHeight Then
                    Dim iVideoHeight As Integer = iControlHeight
                    Dim iVideoWidth As Integer = iControlHeight * camResolution.Height / camResolution.Width
                    Dim iContentWidth As Integer = panel1.Width - panel1.Margin.Left - panel1.Margin.Right - panel1.Padding.Left - panel1.Padding.Right
                    pictureBox1.Location = New Point((iContentWidth - iVideoWidth) / 2, 0)
                    pictureBox1.Size = New Size(iVideoWidth, iVideoHeight)
                Else
                    Dim iVideoWidth As Integer = iControlWidth
                    Dim iVideoHeight As Integer = iControlWidth * camResolution.Width / camResolution.Height
                    Dim iContentHeight As Integer = panel1.Height - panel1.Margin.Top - panel1.Margin.Bottom - panel1.Padding.Top - panel1.Padding.Bottom
                    pictureBox1.Location = New Point(0, (iContentHeight - iVideoHeight) / 2)
                    pictureBox1.Size = New Size(iVideoWidth, iVideoHeight)
                End If
            End If
        End If
    End Sub
    Private Sub ResizePictureBox()
        If m_CurrentCamera IsNot Nothing Then
            Dim camResolution As CamResolution = m_CurrentCamera.CurrentResolution

            If camResolution IsNot Nothing AndAlso camResolution.Width > 0 AndAlso camResolution.Height > 0 Then

                If True Then
                    Dim iVideoWidth As Integer = iControlWidth
                    Dim iVideoHeight As Integer = iControlWidth * camResolution.Height / camResolution.Width
                    Dim iContentHeight As Integer = panel1.Height - panel1.Margin.Top - panel1.Margin.Bottom - panel1.Padding.Top - panel1.Padding.Bottom

                    If iVideoHeight < iContentHeight Then
                        pictureBox1.Location = New Point(0, (iContentHeight - iVideoHeight) / 2)
                        pictureBox1.Size = New Size(iVideoWidth, iVideoHeight)
                    Else
                        pictureBox1.Location = New Point(0, 0)
                        pictureBox1.Size = New Size(pictureBox1.Width, pictureBox1.Height)
                    End If
                End If
            End If
        End If
    End Sub
    Delegate Function OnRefreshImageHandler(ByVal bit As Bitmap)
    Private Function OnRefreshImage(ByVal bit As Bitmap)
        pictureBox1.Image = bit
    End Function
    Dim mRefreshImage As New OnRefreshImageHandler(AddressOf OnRefreshImage)
    Private Sub SetPicture(img As Image)
        Dim temp As Bitmap = (CType((img), Bitmap)).Clone(New Rectangle(0, 0, img.Width, img.Height), img.PixelFormat)

        If pictureBox1.InvokeRequired Then
            pictureBox1.BeginInvoke(mRefreshImage, temp)
        Else
            pictureBox1.Image = temp
        End If
    End Sub
    Private Sub btnCaptureImage_Click(sender As Object, e As EventArgs) Handles btnCaptureImage.Click
        If m_CurrentCamera IsNot Nothing Then
            Dim tempBmp As Bitmap = m_CurrentCamera.GrabImage()
            m_ImageCore.IO.LoadImage(tempBmp)
        End If

    End Sub
    Private Sub m_CurrentCamera_OnFrameCaptrue(bitmap As Bitmap)
        SetPicture(bitmap)
    End Sub
    Private Sub picbox90_MouseHover(sender As Object, e As EventArgs) Handles picbox90.MouseHover
        picbox90.Image = Global.WebcamDemo.Resources._90_hover
    End Sub

    Private Sub picbox90_MouseLeave(sender As Object, e As EventArgs) Handles picbox90.MouseLeave
        picbox90.Image = Global.WebcamDemo.Resources._90_normal
    End Sub

    Private Sub picbox180_MouseHover(sender As Object, e As EventArgs) Handles picbox180.MouseHover
        picbox180.Image = Global.WebcamDemo.Resources._180_hover
    End Sub

    Private Sub picbox180_MouseLeave(sender As Object, e As EventArgs) Handles picbox180.MouseLeave
        picbox180.Image = Global.WebcamDemo.Resources._180_normal
    End Sub
    Private Sub picbox270_MouseHover(sender As Object, e As EventArgs) Handles picbox270.MouseHover
        picbox270.Image = Global.WebcamDemo.Resources._270_hover
    End Sub
    Private Sub picbox270_MouseLeave(sender As Object, e As EventArgs) Handles picbox270.MouseLeave
        picbox270.Image = Global.WebcamDemo.Resources._270_normal
    End Sub
    Private Sub picbox90_Click(sender As Object, e As EventArgs) Handles picbox90.Click
        m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_90)
        RotatePictureBox()
    End Sub

    Private Sub picbox180_Click(sender As Object, e As EventArgs) Handles picbox180.Click
        m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_180)
        ResizePictureBox()
    End Sub

    Private Sub picbox270_Click(sender As Object, e As EventArgs) Handles picbox270.Click
        m_CurrentCamera.RotateVideo(Dynamsoft.UVC.Enums.EnumVideoRotateType.Rotate_270)
        RotatePictureBox()
    End Sub
End Class
